docs/src/writing-constructs.rst

```diff
@@ -0,0 +1,208 @@
+  part of constructing this construct. For example, all resources that require a
+  role allow passing in a **Role** object (specifically, a **RoleRef** object),
+  but if the user does not supply one an appropriate **Role** object is
+  constructed in-place.
+* Use contextual defaulting between properties; the value of one property may
+  affect sensible defaults for other properties. *Example*:
+  **enableDnsHostnames**
+  and and **enableDnsSupport**. **dnsHostnames** requires **dnsSupport**, only
+  yell if the user has *explicitly* disabled DNS Support, but tried to enable
+  DNS Hostnames.  A user expects things to just work - make that the case unless
+  they tell us otherwise.
+* Make the user thinkg about concepts, not implementation detail; for example,
+  if establishing an association between two resources (such as a **Topic**
+  and a **Queue**) requires multiple steps (in this case, creating a
+  **Subscription** but also setting appropriate IAM permissions), make
+  both things happen in a single call (to **subscribeQueue()**). The user
+  should be mentally thinking about the result they're trying to achieve,
+  not the various steps necessary to get there.
+* Make sure you are not introducing new concepts or terminology. For example
+  don't be temped to replace SQS's **dataKeyReusePeriod** with **keyRotation**
+  because it will be hard for people to diagnose problems. They won't be able to
+  just search for *sqs dataKeyReuse* and find topics on it.
+* Indicate units of measurement in property names that don't use a strong type.
+  Use **milli**, **sec**, **min**, **hr**, **Bytes**, **KiB**, **MiB**, **GiB**,
+  etc.
+* Optimize for the common case; example: **Fleet** accepts a **VPC** and deploys
+  in the private subnet by default because that's the common case, but has an
+  option to **placementOptions** for the special case.
+* Use Typescript; in order for construct libraries to be reusable across
+  programming languages (using jsii), they need to be authored in a language
+  that can compile to a jsii assembly. At the moment, the only supported
+  language is TypeScript.
+* Be sure to define an abstract **XxxRef** class for your resources and
+  implement **export()** and **import()** functions; these make it possible
+  to interoperate with resources that are not defined in the same CDK app
+  (they may be manually created, created using raw CloudFormation, or created
+  in a completely unrelated CDK app).
+
+Implementation Details
+======================
+
+* Every construct consists of an exported class (**MyConstruct**) and an
+  exported interface (**MyConstructProps**) that defines the parameters for
+  these classes. THe props argument is the 3rd to the construct (after the
+  mandatory **parent** and **id** arguments), and the entire parameter should be
+  optional if all of the properties on the props object are optional.
+* Constructs should be immutable (as much as possible); sometimes we use
```

Comment:
I don't think that's a reasonable guidance. Many of our constructs are not mutable and in fact, I think in many cases it's useful to have a set of mutation methods that complement props (in most cases, this is relevant for arrays), such as `addXxx`. 

I think the guidance is that the class is responsible to make sure that at any given point in time, it is fully initialized and maintains it's invariants.